### ISSUE TITLE
[ITensorGPU] [BUG] DMRG not working with Observer

### ISSUE THREAD

**Issue by VietTralala** [ITensor/ITensors.jl#943]
**Description of bug**

without a DMRGObserver object my code works on the GPU, however I want to early stop depending on some tolerance on the energy. For this I used the DMRGObserver successfully on the CPU, however when I try to combine the DMRGObserver with GPU calculations I get an error.

**Minimal code demonstrating the bug or unexpected behavior**

<details><summary>Minimal runnable code</summary><p>

```julia
using ITensors
using ITensorGPU
using Random

# Set to identity to run on CPU
gpu = cu
# gpu = identity

N = 5
Random.seed!(1234)
sites = siteinds("Qubit", N; conserve_qns=false)

ampo = AutoMPO()
for j in 1:(N - 1)
ampo .+= 0.5, "S+", j, "S-", j + 1
ampo .+= 0.5, "S-", j, "S+", j + 1
ampo .+= "Sz", j, "Sz", j + 1
end
H = gpu(MPO(ampo, sites))


psi0 = randomMPS(sites)
psi0 = gpu(psi0)

sweeps = Sweeps(10)
maxdim!(sweeps, 10,10,20,20,50,100,200,200,500,1000,1500)
mindim!(sweeps, 1, 10)
cutoff!(sweeps, 1e-11)
noise!(sweeps, 1e-10)
# without dmrg observer it works
# energy, ψ = @time dmrg(H, psi0, sweeps)

# with dmrg observer it doesnt
observer = DMRGObserver(["Z"], sites, energy_tol=1e-4, minsweeps=10)
E, ψ = @time dmrg(H, psi0, sweeps; observer=observer, outputlevel=1)
@show energy
```
</p></details>

**Expected output or behavior**

doing some sweeps and print the energy

**Actual output or behavior**

No sweep is performed. Only error is printed
<details><summary>Output of minimal runnable code</summary><p>

```julia
ERROR: MethodError: no method matching similartype(::Type{DenseVector{Float64}}, ::Type{Float64})
Closest candidates are:
  similartype(::Type{ArrayT}, ::Type) where ArrayT<:(Union{Adjoint{T, AW}, Diagonal{T, AW}, Hermitian{T, AW}, LowerTriangular{T, AW}, Symmetric{T, AW}, Transpose{T, AW}, UnitLowerTriangular{T, AW}, UnitUpperTriangular{T, AW}, UpperTriangular{T, AW}, Base.ReshapedArray{T, var"#s4", AW, MI} where {var"#s4", MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, N} where N}}, SubArray{T, var"#s5", AW, I, L} where {var"#s5", I, L}} where {T, AW}) at /system/user/tran/.julia/packages/NDTensors/lbVmG/src/similar.jl:55
  similartype(::Type{ArrayT}, ::Type, ::Tuple) where ArrayT<:(Union{Adjoint{T, AW}, Diagonal{T, AW}, Hermitian{T, AW}, LowerTriangular{T, AW}, Symmetric{T, AW}, Transpose{T, AW}, UnitLowerTriangular{T, AW}, UnitUpperTriangular{T, AW}, UpperTriangular{T, AW}, Base.ReshapedArray{T, var"#s4", AW, MI} where {var"#s4", MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, N} where N}}, SubArray{T, var"#s5", AW, I, L} where {var"#s5", I, L}} where {T, AW}) at /system/user/tran/.julia/packages/NDTensors/lbVmG/src/similar.jl:51
  similartype(::Type{var"#s68"} where var"#s68"<:(NDTensors.BlockSparseTensor{ElT, NT, var"#s66", var"#s67"} where {var"#s67"<:(NDTensors.BlockSparse{ElT, VecT, N} where N), var"#s66"}), ::Type{IndsR}) where {NT, ElT, VecT, IndsR} at /system/user/tran/.julia/packages/NDTensors/lbVmG/src/blocksparse/blocksparsetensor.jl:13
  ...
Stacktrace:
  [1] similar(#unused#::Type{DenseVector{Float64}}, #unused#::Type{Float64}, size::Int64)
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/similar.jl:35
  [2] similar(#unused#::Type{NDTensors.Dense{Float64, DenseVector{Float64}}}, #unused#::Type{Float64}, length::Int64)
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/dense.jl:93
  [3] _similar_from_dims(#unused#::Type{NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, DenseVector{Float64}}}}, #unused#::Type{Float64}, dims::Tuple{Index{Int64}, Index{Int64}, Index{Int64}})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/tensor.jl:202
  [4] _similar_from_dims(#unused#::Type{NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, DenseVector{Float64}}}}, dims::Tuple{Index{Int64}, Index{Int64}, Index{Int64}})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/tensor.jl:196
  [5] similar(#unused#::Type{NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, DenseVector{Float64}}}}, dims::
Tuple{Index{Int64}, Index{Int64}, Index{Int64}})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/tensor.jl:169
  [6] contraction_output(#unused#::NDTensors.DenseTensor{Float64, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, Vector{Float64}}}, #unused#::NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}}, indsR::Tuple{Index{Int64}, Index{Int64}, Index{Int64}})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/dense.jl:598
  [7] contraction_output(T1::NDTensors.DenseTensor{Float64, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, Vector{Float64}}}, labelsT1::Tuple{Int64, Int64}, T2::NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}}, labelsT2::Tuple{Int64, Int64, Int64}, labelsR::Tuple{Int64, Int64, Int64})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/dense.jl:604
  [8] contract(T1::NDTensors.DenseTensor{Float64, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, Vector{Float64}}}, labelsT1::Tuple{Int64, Int64}, T2::NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}}, labelsT2::
Tuple{Int64, Int64, Int64}, labelsR::Tuple{Int64, Int64, Int64})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/dense.jl:620
  [9] contract(T1::NDTensors.DenseTensor{Float64, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, Vector{Float64}}}, labelsT1::Tuple{Int64, Int64}, T2::NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}}, labelsT2::Tuple{Int64, Int64, Int64})
    @ NDTensors ~/.julia/packages/NDTensors/lbVmG/src/dense.jl:611
 [10] _contract(A::NDTensors.DenseTensor{Float64, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, Vector{Float64}}}, B::NDTensors.DenseTensor{Float64, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float64, CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}}})
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/itensor.jl:1851
 [11] _contract(A::ITensor, B::ITensor)
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/itensor.jl:1857
 [12] contract(A::ITensor, B::ITensor)
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/itensor.jl:1961
 [13] *(A::ITensor, B::ITensor)
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/itensor.jl:1948
 [14] measurelocalops!(obs::DMRGObserver{Float64}, wf::ITensor, i::Int64)
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/mps/observer.jl:137
 [15] measure!(obs::DMRGObserver{Float64}; kwargs::Base.Iterators.Pairs{Symbol, Any, NTuple{8, Symbol}, NamedTuple{(:energy, :psi, :bond, :sweep, :half_sweep, :spec, :outputlevel, :sweep_is_done), Tuple{Float64, MPS, Int64, Int64, Int64, Spectrum{CUDA.CuArray{Float64, 1, CUDA.Mem.DeviceBuffer}, Float64}, Int64, Bool}}})
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/mps/observer.jl:165
 [16] macro expansion
    @ ~/.julia/environments/v1.6/dev/ITensors/src/mps/dmrg.jl:285 [inlined]
 [17] macro expansion
    @ ./timing.jl:287 [inlined]
 [18] dmrg(PH::ProjMPO, psi0::MPS, sweeps::Sweeps; kwargs::Base.Iterators.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:observer, :outputlevel), Tuple{DMRGObserver{Float64}, Int64}}})
    @ ITensors ~/.julia/environments/v1.6/dev/ITensors/src/mps/dmrg.jl:186
 [19] #dmrg#961
    @ ~/.julia/environments/v1.6/dev/ITensors/src/mps/dmrg.jl:47 [inlined]
 [20] macro expansion
    @ ./timing.jl:210 [inlined]
 [21] test_gpu_1d()
    @ Main ~/zelle/provable-ml-quantum/julia_scripts/test_gpu.jl:95
 [22] top-level scope
    @ REPL[135]:1
 [23] top-level scope
    @ ~/.julia/packages/CUDA/5jdFl/src/initialization.jl:52

```
</p></details>

**Version information**

 - Output from `versioninfo()`:
```julia
julia> versioninfo()
Julia Version 1.6.6
Commit b8708f954a (2022-03-28 07:17 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2687W v4 @ 3.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, broadwell)

```
 - Output from `using Pkg; Pkg.status("ITensors")`:
```julia
julia> using Pkg; Pkg.status("ITensors")
      Status `~/.julia/environments/v1.6/Project.toml`
  [9136182c] ITensors v0.3.15 `dev/ITensors`

julia> using Pkg; Pkg.status("ITensorGPU")
      Status `~/.julia/environments/v1.6/Project.toml`
  [d89171c1] ITensorGPU v0.0.5
```

ITensors was installed locally in develop mode, however the code wasn't changed

**Comment by mtfishman** [comment]
Thanks for the report @VietTralala.

I think the issue is that the observer is creating an operator internally which is by default being allocated on CPU, and contracting an ITensor allocated on CPU with one allocated on GPU is not allowed. We will have to dig into the code and make sure the operator gets allocated properly on GPU when needed.

Something you could as a workaround is define your own Observer (following the documentation here: https://itensor.github.io/ITensors.jl/stable/Observer.html#Defining-a-Custom-Observer) where you compute the expectation value yourself, properly allocating the ITensors on GPU.

**Comment by mtfishman** [comment]
This is fixed by #956.
